subprojects/core/src/main/java/org/gradle/caching/internal/controller/DefaultBuildCacheController.java

```diff
@@ -17,108 +17,236 @@
+                public void run(BuildOperationContext context) {
+                    try {
+                        result = command.load(new FileInputStream(file));
+                    } catch (FileNotFoundException e) {
+                        throw UncheckedException.throwAsUncheckedException(e);
+                    }
+
+                    context.setResult(new UnpackOperationResult(
+                        result.getArtifactEntryCount()
+                    ));
+                }
+
+                @Override
+                public BuildOperationDescriptor.Builder description() {
+                    return BuildOperationDescriptor.displayName("Unpack " + command.getKey() + " build cache entry")
+                        .details(new UnpackOperationDetails(command.getKey(), file.length()));
+                }
+            });
         }
+
     }
 
-    private void doStoreBoth(BuildCacheStoreCommand command) {
-        File destination = temporaryFileProvider.createTemporaryFile("gradle_cache", "entry");
-        try {
-            BuildCacheStoreCommand.Result result = doTmpFileStore(command, destination);
-            local.doStore(command.getKey(), destination, result);
-            remote.doStore(command.getKey(), destination, result);
-        } catch (IOException e) {
-            throw new UncheckedIOException(e);
-        } finally {
-            GFileUtils.deleteQuietly(destination);
+    @Override
+    public void store(final BuildCacheStoreCommand command) {
+        boolean anyStore = local.canStore() || legacyLocal.canStore() || remote.canStore();
+        if (!anyStore) {
+            return;
         }
+
+        final BuildCacheKey key = command.getKey();
+        final Pack pack = new Pack(command);
+
+        tmp.allocate(command.getKey(), new Action<File>() {
+            @Override
+            public void execute(File file) {
+                pack.execute(file);
+
+                if (legacyLocal.canStore()) {
+                    legacyLocal.store(key, new StoreTarget(file));
+                }
+
+                if (remote.canStore()) {
+                    remote.store(key, new StoreTarget(file));
+                }
+
+                if (local.canStore()) {
+                    local.store(key, file);
+                }
+            }
+        });
     }
 
-    private BuildCacheStoreCommand.Result doTmpFileStore(BuildCacheStoreCommand storeOp, File destination) throws IOException {
-        OutputStream fileOutputStream = null;
-        try {
-            fileOutputStream = new BufferedOutputStream(new FileOutputStream(destination));
-            return storeOp.store(fileOutputStream);
-        } catch (FileNotFoundException e) {
-            throw new BuildCacheException("Couldn't create local file for cache entry", e);
-        } finally {
-            IOUtils.closeQuietly(fileOutputStream);
+    private class Pack implements Action<File> {
+
+        private final BuildCacheStoreCommand command;
```

Comment:
From the controller client's PoV, it's a store operation. I think this name shouldn't be changed.